File: composeApp/src/desktopMain/kotlin/org/queststudios/yamlvalidation/core/ValidatorCore.kt
package org.queststudios.yamlvalidation.core

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import org.queststudios.yamlvalidation.validation.ValidationLogger
import org.queststudios.yamlvalidation.validation.ValidationRule
import java.io.File
import java.io.FileWriter
import java.io.IOException
import org.queststudios.yamlvalidation.i18n.Strings
import org.yaml.snakeyaml.Yaml
import org.yaml.snakeyaml.constructor.SafeConstructor
import org.yaml.snakeyaml.LoaderOptions

data class ValidationResult(val success: Boolean, val errorMessage: String? = null)

class ValidatorCore(
    private val yamlPath: String,
    private val logger: ValidationLogger,
    private val rules: List<ValidationRule>,
    private val logFilePath: String? = null,
    private val language: String = "es", // Ahora configurable
    private val spectralPath: String? = null // NUEVO: ruta para spectral
) {
    var yamlData: Map<String, Any?> = emptyMap()
        private set
    private val logHistory = mutableListOf<String>()

    private fun logToFile(level: String, message: String) {
        logFilePath?.let {
            try {
                FileWriter(it, true).use { fw ->
                    fw.write("[$level] $message\n")
                }
            } catch (e: IOException) {
                val errMsg = Strings.get(language, "validation_log_file_write_error") + ": ${e.message}"
                logger.log("ERROR", errMsg)
                logHistory.add("[ERROR] $errMsg")
            }
        }
    }

    private fun log(level: String, message: String) {
        logger.log(level, message)
        logToFile(level, message)
        logHistory.add("[$level] $message")
    }

    fun getLogHistory(): List<String> = logHistory.toList()

    fun loadYaml(): ValidationResult {
        if (yamlPath.isBlank()) {
            val msg = Strings.get(language, "error.noYaml")
            log("ERROR", msg)
            yamlData = emptyMap()
            return ValidationResult(false, msg)
        }
        try {
            val input = File(yamlPath)
            if (!input.exists() || !input.isFile) {
                val msg = Strings.get(language, "validation_yaml_file_not_found")
                log("ERROR", msg)
                yamlData = emptyMap()
                return ValidationResult(false, msg)
            }
            val yaml = Yaml(SafeConstructor(LoaderOptions()))
            val loaded = input.inputStream().use { yaml.load<Map<String, Any?>>(it) }
            yamlData = loaded ?: emptyMap()
            if (!yamlData.containsKey("paths")) {
                val msg = Strings.get(language, "validation_yaml_missing_paths")
                log("ERROR", msg)
                yamlData = emptyMap()
                return ValidationResult(false, msg)
            }
            // Log de éxito tras carga
            log("SUCCESS", "YAML loaded successfully. Keys: ${yamlData.keys}")
        } catch (e: Exception) {
            val msg = Strings.get(language, "validation_yaml_load_error") + ": ${e.message ?: ""}\n${e.stackTraceToString()}"
            log("ERROR", msg)
            yamlData = emptyMap()
            return ValidationResult(false, msg)
        }
        return ValidationResult(true)
    }

    data class ParallelValidationResult(
        val internal: ValidationResult,
        val spectral: String // salida de spectral (si hay error, también aquí)
    )

    suspend fun runAllValidationsParallel(): ParallelValidationResult = coroutineScope {
        val spectralDeferred = async(Dispatchers.IO) { runSpectralValidationOutput() }
        val internalDeferred = async(Dispatchers.Default) { runInternalValidations() }
        ParallelValidationResult(
            internal = internalDeferred.await(),
            spectral = spectralDeferred.await()
        )
    }

    private fun runSpectralValidationOutput(): String {
        // 1. Leer el YAML y modificar el campo 'info.title' temporalmente
        val yamlFile = File(yamlPath)
        val yamlText = yamlFile.readText()
        val yaml = org.yaml.snakeyaml.Yaml()
        val yamlData: MutableMap<String, Any?> = yaml.load(yamlText) as? MutableMap<String, Any?> ?: return "No se pudo leer el YAML."
        val info = yamlData["info"] as? MutableMap<String, Any?>
        val originalTitle = info?.get("title")?.toString()
        if (info != null) {
            info["title"] = "TemporalSpectralName"
        }
        // 2. Guardar el YAML modificado en un archivo temporal
        val tempFile = File.createTempFile("spectral_temp", ".yaml")
        tempFile.writeText(org.yaml.snakeyaml.Yaml().dump(yamlData))
        try {
            val spectralCmd = listOf(
                "cmd", "/c",
                "spectral lint -r ./poc/.spectral_v2.yaml -f pretty \"${tempFile.absolutePath}\""
            )
            val processBuilder = ProcessBuilder(spectralCmd)
            val spectralDir = spectralPath?.takeIf { it.isNotBlank() }?.let { File(it) } ?: File(System.getProperty("user.dir"))
            processBuilder.directory(spectralDir)
            processBuilder.redirectErrorStream(true)
            val process = processBuilder.start()
            val output = process.inputStream.bufferedReader().readText()
            process.waitFor()
            process.destroy()
            return output
        } catch (e: Exception) {
            return "Error ejecutando Spectral: ${e.message}\n${e.stackTraceToString()}"
        } finally {
            tempFile.delete()
        }
    }

    private fun runInternalValidations(): ValidationResult {
        val loadResult = loadYaml()
        if (!loadResult.success) return loadResult
        val paths = yamlData["paths"] as? Map<*, *> ?: emptyMap<Any?, Any?>()
        for (endpoint in paths.keys) {
            val endpointObj = paths[endpoint] as? Map<*, *> ?: continue
            for (method in endpointObj.keys) {
                val methodObj = endpointObj[method] as? Map<*, *> ?: continue
                val context = ValidationContext(yamlData)
                for (rule in rules) {
                    rule.validate(endpoint.toString(), method.toString(), context, object : ValidationLogger {
                        override fun log(level: String, message: String) {
                            this@ValidatorCore.log(level, message)
                        }
                    })
                }
            }
        }
        return ValidationResult(true)
    }

    fun exportSpectralToFile(outputPath: String): ValidationResult {
        try {
            val spectralCmd = listOf(
                "cmd", "/c",
                "spectral lint -r ./poc/.spectral_v2.yaml -f pretty \"$yamlPath\""
            )
            val spectralDir = spectralPath?.takeIf { it.isNotBlank() }?.let { File(it) } ?: File(System.getProperty("user.dir"))
            val processBuilder = ProcessBuilder(spectralCmd)
            processBuilder.directory(spectralDir)
            processBuilder.redirectErrorStream(true)
            val process = processBuilder.start()
            val output = process.inputStream.bufferedReader().readText()
            process.waitFor()
            logger.log("SPECTRAL", output)
            logToFile("SPECTRAL", output)

            // Escribir la salida al fichero de destino
            File(outputPath).writeText(output)

            process.destroy()
            return ValidationResult(true)
        } catch (e: Exception) {
            val msg = "Error ejecutando Spectral export: ${e.message}\n${e.stackTraceToString()}"
            logger.log("ERROR", msg)
            logToFile("ERROR", msg)
            return ValidationResult(false, msg)
        }
    }
}

File: src/core/ValidatorCore.js
import { Strings } from '../i18n/strings.js';

// Use yaml from window (exposed by preload)
const yaml = window.yaml;

export class ValidationResult {
  constructor(success, errorMessage = null) {
    this.success = success;
    this.errorMessage = errorMessage;
  }
}

export class ValidatorCore {
  constructor(yamlPath, logger, rules, logFilePath = null, language = 'es', spectralPath = null) {
    this.yamlPath = yamlPath;
    this.logger = logger;
    this.rules = rules;
    this.logFilePath = logFilePath;
    this.language = language;
    this.spectralPath = spectralPath;
    this.yamlData = {};
    this.logHistory = [];
  }

  getLogHistory() {
    return [...this.logHistory];
  }

  async loadYaml() {
    if (!this.yamlPath || this.yamlPath.trim() === '') {
      const msg = Strings.get(this.language, 'error.noYaml');
      this.log('ERROR', msg);
      this.yamlData = {};
      return new ValidationResult(false, msg);
    }

    try {
      const result = await window.electronAPI.readFile(this.yamlPath);
      if (!result.success) {
        const msg = Strings.get(this.language, 'validation_yaml_file_not_found');
        this.log('ERROR', msg);
        this.yamlData = {};
        return new ValidationResult(false, msg);
      }

      this.yamlData = yaml.load(result.content);
      
      if (!this.yamlData || !this.yamlData.paths) {
        const msg = Strings.get(this.language, 'validation_yaml_missing_paths');
        this.log('ERROR', msg);
        this.yamlData = {};
        return new ValidationResult(false, msg);
      }

      this.log('SUCCESS', `YAML loaded successfully. Keys: ${Object.keys(this.yamlData).join(', ')}`);
      return new ValidationResult(true);
    } catch (error) {
      const msg = `${Strings.get(this.language, 'validation_yaml_load_error')}: ${error.message}`;
      this.log('ERROR', msg);
      this.yamlData = {};
      return new ValidationResult(false, msg);
    }
  }

  async runAllValidationsParallel() {
    // Solo ejecutar Spectral si hay spectralPath (solo para licencias de NTT)
    if (this.spectralPath) {
      const [spectralResult, internalResult] = await Promise.all([
        this.runSpectralValidationOutput(),
        this.runInternalValidations()
      ]);

      return {
        internal: internalResult,
        spectral: spectralResult
      };
    } else {
      // Si no hay spectralPath, solo ejecutar validaciones internas
      const internalResult = await this.runInternalValidations();
      return {
        internal: internalResult,
        spectral: ''
      };
    }
  }

  async runSpectralValidationOutput() {
    try {
      // Read YAML file
      const result = await window.electronAPI.readFile(this.yamlPath);
      if (!result.success) {
        return 'No se pudo leer el YAML.';
      }

      const yamlData = yaml.load(result.content);
      
      // Modify title temporarily
      if (yamlData.info) {
        yamlData.info.title = 'TemporalSpectralName';
      }

      // Create temporary file
      const tempYaml = yaml.dump(yamlData);
      const tempPath = `${this.yamlPath}.spectral_temp.yaml`;
      
      await window.electronAPI.writeFile(tempPath, tempYaml);

      // Run Spectral
      const spectralResult = await window.electronAPI.runSpectral({
        spectralPath: this.spectralPath,
        yamlPath: tempPath,
        spectralDir: this.spectralPath || undefined
      });

      // Note: Temp file cleanup would need to be handled by the main process
      // For now, we'll leave it for manual cleanup or implement a cleanup handler later

      return spectralResult.success ? spectralResult.output : `Error ejecutando Spectral: ${spectralResult.error}`;
    } catch (error) {
      return `Error ejecutando Spectral: ${error.message}`;
    }
  }

  async runInternalValidations() {
    const loadResult = await this.loadYaml();
    if (!loadResult.success) {
      return loadResult;
    }

    const paths = this.yamlData.paths || {};
    
    for (const [endpoint, endpointObj] of Object.entries(paths)) {
      if (!endpointObj || typeof endpointObj !== 'object') continue;
      
      for (const [method, methodObj] of Object.entries(endpointObj)) {
        if (!methodObj || typeof methodObj !== 'object') continue;
        
        const context = new ValidationContext(this.yamlData);
        
        for (const rule of this.rules) {
          rule.validate(endpoint, method, context, {
            log: (level, message) => this.log(level, message)
          });
        }
      }
    }

    return new ValidationResult(true);
  }

  async exportSpectralToFile(outputPath) {
    try {
      const result = await window.electronAPI.runSpectral({
        spectralPath: this.spectralPath,
        yamlPath: this.yamlPath,
        spectralDir: this.spectralPath || undefined
      });

      if (result.success) {
        await window.electronAPI.writeFile(outputPath, result.output);
        this.log('SPECTRAL', result.output);
        return new ValidationResult(true);
      } else {
        const msg = `Error ejecutando Spectral export: ${result.error}`;
        this.log('ERROR', msg);
        return new ValidationResult(false, msg);
      }
    } catch (error) {
      const msg = `Error ejecutando Spectral export: ${error.message}`;
      this.log('ERROR', msg);
      return new ValidationResult(false, msg);
    }
  }

  log(level, message) {
    this.logger.log(level, message);
    this.logHistory.push(`[${level}] ${message}`);
    
    if (this.logFilePath) {
      // Note: File logging would need to be done via IPC in Electron
      // For now, we just log to console
      console.log(`[${level}] ${message}`);
    }
  }
}

export class ValidationContext {
  constructor(yamlData) {
    this.yamlData = yamlData;
  }
}

